x-pack/filebeat/input/cometd/input.go

```diff
@@ -0,0 +1,204 @@
+			}
+		}
+	}
+	return nil
+}
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(errors.Wrapf(err, "failed to register %v input", inputName))
+	}
+}
+
+// NewInput creates a new CometD input that consumes events from
+// a topic subscription.
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (inp input.Input, err error) {
+	// Extract and validate the input's configuration.
+	conf := defaultConfig()
+	if err = cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	os.Setenv("SALESFORCE_CONSUMER_KEY", conf.Auth.OAuth2.ClientID)
+	os.Setenv("SALESFORCE_CONSUMER_SECRET", conf.Auth.OAuth2.ClientSecret)
+	os.Setenv("SALESFORCE_USER", conf.Auth.OAuth2.User)
+	os.Setenv("SALESFORCE_PASSWORD", conf.Auth.OAuth2.Password)
+	os.Setenv("SALESFORCE_TOKEN_URL", conf.Auth.OAuth2.TokenURL)
+
+	logger := logp.NewLogger("cometd").With(
+		"pubsub_channel", conf.ChannelName)
+
+	// Wrap input.Context's Done channel with a context.Context. This goroutine
+	// stops with the parent closes the Done channel.
+	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+	go func() {
+		defer cancelInputCtx()
+		select {
+		case <-inputContext.Done:
+		case <-inputCtx.Done():
+		}
+	}()
+
+	// If the input ever needs to be made restartable, then context would need
+	// to be recreated with each restart.
+	workerCtx, workerCancel := context.WithCancel(inputCtx)
+
+	in := &cometdInput{
+		config:       conf,
+		log:          logger,
+		inputCtx:     inputCtx,
+		workerCtx:    workerCtx,
+		workerCancel: workerCancel,
+		ackedCount:   atomic.NewUint32(0),
+	}
+
+	// Creating a new channel for cometd input
+	in.out = make(chan bay.TriggerEvent)
+
+	// Build outlet for events.
+	in.outlet, err = connector.Connect(cfg)
+	if err != nil {
+		return nil, err
+	}
+	in.log.Infof("Initialized %s input.", inputName)
+	return in, nil
+}
+
+// Stop stops the input and waits for it to fully stop.
+func (in *cometdInput) Stop() {
+	close(in.out)
+	in.workerCancel()
+	in.workerWg.Wait()
+}
+
+// Wait is an alias for Stop.
+func (in *cometdInput) Wait() {
+	in.Stop()
+}
+
+type cometdInput struct {
+	config
+
+	log      *logp.Logger
+	outlet   channel.Outleter // Output of received messages.
+	inputCtx context.Context  // Wraps the Done channel from parent input.Context.
+
+	workerCtx    context.Context    // Worker goroutine context. It's cancelled when the input stops or the worker exits.
+	workerCancel context.CancelFunc // Used to signal that the worker should stop.
+	workerOnce   sync.Once          // Guarantees that the worker goroutine is only started once.
+	workerWg     sync.WaitGroup     // Waits on worker goroutine.
+
+	ackedCount *atomic.Uint32                   // Total number of successfully ACKed messages.
+	Transport  httpcommon.HTTPTransportSettings `config:",inline"`
+	Retry      retryConfig                      `config:"retry"`
```

Comment:
just curious: why are these properties exported?

Reply:
We don't require them now as we are leveraging Bayeux library.

Reply:
Same as [here](https://github.com/elastic/beats/pull/30089#discussion_r840477198). We should open a PR to the fork and polish that.